### PR TITLE
[DataGridPremium] Fix crash when exporting all columns to Excel

### LINKED ISSUE #5353: [data grid] ColSpan undefined if allColumns: true Excel Export
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

If you have more than one column hide and `allColumns: true` in excel options there is an error:
image

CSB: https://codesandbox.io/s/excelcustomexport-demo-mui-x-forked-lmqsv7?file=/demo.tsx

### Expected behavior 🤔

Export should work with hiding columns

### Steps to reproduce 🕹

Steps:

1. Hide more than one column
2. Pass excel options with `allColumns: true`
3. Press Excel Export

### Context 🔦

_No response_

### Your environment 🌎

_No response_

### Order ID 💳 (optional)

_No response_

### PR DESCRIPTION
**Pull request by cherniavskii**
Fixes #5353

Codesandbox: https://codesandbox.io/s/excelcustomexport-demo-mui-x-forked-qt3tw0

### CODE REVIEW

File: packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnSpanning.ts [hunk 1]
```diff
@@ -44,6 +44,12 @@ export const useGridColumnSpanning = (apiRef: React.MutableRefObject<GridApiComm
       const columnsLength = visibleColumns.length;
       const column = visibleColumns[columnIndex];
```
Comment (alexfauquette) [inline packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnSpanning.ts:45]: What is preventing from adding a parameter `includeHiddenColumns` to the method such that we can do the following?
```suggestion
      const column = includeHiddenColumns ? hiddenColumns[columnIndex] : visibleColumns[columnIndex];
```

File: packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnSpanning.ts [hunk 2]
```diff
@@ -38,9 +38,19 @@ export const useGridColumnSpanning = (apiRef: React.MutableRefObject<GridApiComm
...
-      const visibleColumns = apiRef.current.getVisibleColumns();
+      const {
+        columnIndex,
+        rowId,
+        minFirstColumnIndex,
+        maxLastColumnIndex,
+        includeHiddenColumns = false,
+      } = params;
+
+      const visibleColumns = includeHiddenColumns
```
Comment (m4theushw) [inline packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnSpanning.ts:51]: Wouldn't be simpler to pass the columns to use as a param?

In https://github.com/mui/mui-x/blob/29a1f55ba05a77cd7276608c6ff8fc7516300f21/packages/grid/x-data-grid-premium/src/hooks/features/export/useGridExcelExport.tsx#L36 we already calculate which columns to use. We're doing it again inside `apiRef.current.unstable_calculateColSpan`. Both places where this method is called we have the columns nearby.

```diff
diff --git a/packages/grid/x-data-grid-premium/src/hooks/features/export/serializer/excelSerializer.ts b/packages/grid/x-data-grid-premium/src/hooks/features/export/serializer/excelSerializer.ts
index e58808630..8dc2a399f 100644
--- a/packages/grid/x-data-grid-premium/src/hooks/features/export/serializer/excelSerializer.ts
+++ b/packages/grid/x-data-grid-premium/src/hooks/features/export/serializer/excelSerializer.ts
@@ -65,6 +65,7 @@ const serializeRow = (
     rowId: id,
     minFirstColumn: 0,
     maxLastColumn: columns.length - 1,
+    columns,
   });
 
   columns.forEach((column, colIndex) => {
diff --git a/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx b/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx
index cf3ccb353..18a0e5f97 100644
--- a/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx
+++ b/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx
@@ -400,7 +400,12 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
       const row = currentPage.rows[i];
       renderedRows.push(row);
 
-      apiRef.current.unstable_calculateColSpan({ rowId: row.id, minFirstColumn, maxLastColumn });
+      apiRef.current.unstable_calculateColSpan({
+        rowId: row.id,
+        minFirstColumn,
+        maxLastColumn,
+        columns: visibleColumns,
+      });
     }
 
     const [initialFirstColumnToRender, lastColumnToRender] = getRenderableIndexes({
```

Comment (cherniavskii) [reply packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnSpanning.ts]: Thanks, it makes total sense!

**Review by m4theushw — APPROVED** [review]
Simpler to read now. 👌

**Pull request merged**